Flow table changes to support GTPu tunnel

srakula-trex opened the pull request:
This change set has following additions.
a. Adding an RPC to support enabling of GTPU during server starting.
b. Activate the clients based on the subscriber creations ( happens over RPC)
c. Inserting the tunnel instance in to the packet buffer.

srakula-trex commented:
@hhaim , please review these changes.  This is second set of changes that we have made so far . There is going to be one more PR for creating the tunnel instances via RPC .

hhaim commented:
@srakula-trex looks pretty simple 

* You can verify our regression locally see the wiki how to run it on a loop-back setup 
* Try to think what is generic tunnel and what is GTPU specific and name the function correctly. Tuple Generator for example is generic tunnel  


0. Please add short documentation  for how to use it in ASTF and some design changes. I don't see the big picture yet. 

1. Editor configuration **TAB** -> **space (4)** !!

for example  
```
    CClientPool(){
        m_thread_id=0;
        m_rss_thread_id=0;
        m_rss_thread_max=0;
        m_reta_mask=0;
        m_rss_astf_mode=false;
	m_active_clients.clear();  << please align the code and REMOVE TABS and convert them to spaces (4 spaces for 
```

2. Would rename all the gtpu to tunnel as this object is opaque to gtpu and would be valid for **any** tunnel 

so ```set_gtpu_info_for_clients``` ==> ```set_tunnel_info_for_clients```
	
```	
	void CClientPool::set_gtpu_info_for_clients(uint32_t  min_ip,
                                            uint32_t  max_ip,
                                            bool      add,
                                            void      *gtpu)

```

3. Didn't get this code, what do you want to do here, do you want to enable it only on one port? e.g. 1 what would happen in case of multi-ports (8 ports) on which port GTPU would be enabled?
 
```
 for (i=0; i<port_size; i++){
            cg->m_mac_info[i].copy_src(( char *)g_opts->m_mac_addr[i].u.m_mac.src)   ;
            cg->m_mac_info[i].copy_dest(( char *)g_opts->m_mac_addr[i].u.m_mac.dest)  ;
            g_opts->m_mac_addr[i].u.m_mac.is_set = 1;
            if ( CGlobalInfo::m_options.m_enable_gtpu == i)
            {
              g_opts->m_ip_cfg[i].enable_gtp(1);
            }
```    

 
4. This is STF code, do you support STF or just ASTF?? I would remove the STF support

```
int CFlowGenListPerThread::generate_flows_roundrobin(bool *done){
    // round robin

    CFlowGeneratorRecPerThread * cur;
    bool found=false;
    // try current
    int i;
    *done = true;
    for (i=0;i<(int)m_cap_gen.size();i++ ) {
        cur=m_cap_gen[m_cur_template];
        if (!(cur->m_info->m_limit_was_set) ||
            (cur->m_info->m_flowcnt < cur->m_info->m_limit)) {
            *done = false;
            if ( cur->m_policer.update(1.0,m_cur_time_sec) ){
                cur->m_info->m_flowcnt++;
                found=true;
                break;
            }
        }
        inc_current_template();
    }
```

5. Why did you change the type from uint8 to uint16? ```uint16_t idx``` from ```uint8_t```?? 

```
   via_replace_cmd.m_flags |= CMiniVMCmdBase::MIN_VM_V6;
                    for (uint16_t idx=0; idx<6; idx++){
                        via_replace_cmd.m_ip.v6[idx] = CGlobalInfo::m_options.m_src_ipv6[idx];
                        via_replace_cmd.m_ip_via.v6[idx] = CGlobalInfo::m_options.m_src_ipv6[idx];
                    }
```


6. This will not work in the common case ( no gtpu case) as the default ```CGlobalInfo::m_options.m_enable_gtpu == 0xFF``` 
```
    if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() || CGlobalInfo::m_options.m_enable_gtpu == 0xFF)){
        /* we can't generate a flow, there is a limit*/
        return;
    }
```


```CGlobalInfo::m_options.m_enable_gtpu``` better to convert to 

1. is_tunnel_aware()  >> is there any tunnel configured 
2. is_gtpu_tunnel()   ==> is there a GTPU tunnel configured 

7. 

```
   if (ftp->m_tun_handle == NULL && CGlobalInfo::m_options.m_enable_gtpu != 0xFF) {
        if (m_pctx->m_ctx->is_client_side())
         {
             INC_UDP_STAT(m_pctx, m_tg_id, udps_notunnel);
         }
    }
    m->dynfield1[0] = (uint64_t)ftp->m_tun_handle;  >>> we should not write to this location if we are not in tunnel mode
     >> we should optimized for none-tunnel mode 
  ```   

   8. same issue here
   
  ```
static inline int _tcp_build_cpkt(CPerProfileCtx * pctx,
                                  CFlowTemplate *ftp,
                                  struct tcpcb *tp,
                                  uint16_t tcphlen,
                                  CTcpPkt &pkt){
    int len= ftp->m_offset_l4+tcphlen;
    rte_mbuf_t * m;
    m=tp->pktmbuf_alloc(len);
    pkt.m_buf=m;
    CTcpPerThreadCtx * ctx = pctx->m_ctx;

    if (ftp->m_tun_handle == NULL && CGlobalInfo::m_options.m_enable_gtpu != 0xFF)
     {
       /*Check if we need to add any error counter for the special case*/
       if (ctx->is_client_side())
         {
           INC_STAT(pctx, tp->m_flow->m_tg_id, tcps_notunnel);
         }
     }

    pkt.m_buf->dynfield1[0] = (uint64_t)ftp->m_tun_handle;
        
```


9. void *tun_handle, instead of adding it do alloc_flow better to add a function to set it and only in case of tunnel to do so
  

```

CTcpFlow * alloc_flow(CPerProfileCtx * pctx,
                          uint32_t src,
                          uint32_t dst,
                          uint16_t src_port,
                          uint16_t dst_port,
                          uint16_t vlan,
                          bool is_ipv6,
                          void *tun_handle,
                          uint16_t tg_id=0,
                          uint16_t template_id=0);

    CUdpFlow * alloc_flow_udp(CPerProfileCtx * pctx,
                              uint32_t src,
                              uint32_t dst,
                              uint16_t src_port,
                              uint16_t dst_port,
                              uint16_t vlan,
                              bool is_ipv6,
                              void *tun_handle,
                              bool client,
                              uint16_t tg_id=0,
                              uint16_t template_id=0);
```

for example:

```
flow = alloc_flow(pctx,
                          src,
                          dst,
                          src_port,
                          dst_port,
                          vlan,
                          is_ipv6,
                          tg_id
                          template_id)
if (is_tunnel_aware()){
  flow->setTunnelData(tun_handle)
}
```

srakula-trex commented:
@hhaim , please review these changes.  The Regression was failing because of the flow related changes , which we will address in consecutive PRs . But initial framework to support GTP tunnel would remain the same.

hhaim commented:
Hi @srakula-trex 
* There are still code with tabs and spaces. 
![image](https://user-images.githubusercontent.com/13034429/92569738-a6bbfa80-f289-11ea-9610-e55560117e8b.png)
* I don't see any high level explanation? did you write it , why there is a YAML support, is it needed?

Could you answer my comments?I don't know what you have addressed and what you didn't.

hhaim commented:
@srakula-trex one more thing. You can run the regression on a loopback setup (without GTP-U) before submitting  

see here, for running regression 
https://github.com/cisco-system-traffic-generator/trex-core/wiki

srakula-trex commented:
> ooks pretty simple
> 
> * You can verify our regression locally see the wiki how to run it on a loop-back setup
> * Try to think what is generic tunnel and what is GTPU specific and name the function correctly. Tuple Generator for example is generic tunnel
> 
> 1.

1. I have tried to follow the wiki to Run regression , but hitting some errors . Do you have any video links how to setup this ?

srakula-trex commented:
E-mailed you @hhaim about the possible approach that we can take forward to make quick progress on this ?

kvbhaskar commented on src/gtp_tunnel.h:
```diff
@@ -0,0 +1,73 @@
+public:
+	GTPU();
+
+    GTPU( uint32_t teid, uint32_t src_ip, uint32_t dst_ip) ;
```
please fix the indentation

kvbhaskar commented on src/gtp_tunnel.h:
```diff
@@ -0,0 +1,73 @@
+	}
+
+
+	virtual int Prepend(rte_mbuf * m, u_int8_t is_ipv4, u_int16_t);
```
same indentation problems

kvbhaskar commented on src/gtp_tunnel.h:
```diff
@@ -0,0 +1,73 @@
+
+    int Adjust_ipv6_tunnel( rte_mbuf * pkt, struct rte_ether_hdr const *eth);
+
+	virtual int Adjust(rte_mbuf * m, u_int8_t);
```
same indentation issues

kvbhaskar commented on src/gtp_tunnel.h:
```diff
@@ -0,0 +1,73 @@
+
+
+private:
+	uint32_t teid;
```
please fix it

kvbhaskar commented on src/tunnel.h:
```diff
@@ -0,0 +1,38 @@
+	static void InstallRxCallback(uint16_t port, uint16_t queue);
+
+	static void InstallTxCallback(uint16_t port, uint16_t queue);
+    static int Parse(YAML::Node const&, std::vector<Tunnel *> & tunnels);
```
indentation issues

kvbhaskar commented on src/tuple_gen.h:
```diff
@@ -641,22 +681,54 @@ class CClientPool : public CIpPool {
+      CIpInfoBase* ip_info;
+      uint32_t idx;
+
+      if ( CGlobalInfo::m_options.m_enable_gtpu == 0xFF) {
```
method to access please

kvbhaskar commented on src/tuple_gen.h:
```diff
@@ -641,22 +681,54 @@ class CClientPool : public CIpPool {
+	      m_cur_act_itr++;
+          ip_info = m_ip_info[idx];
+          ip_info->generate_tuple(tuple);
+	      tuple.setTunHandle(ip_info->get_tunnel_info());
```
indentation please

kvbhaskar commented on src/tuple_gen.h:
```diff
@@ -641,22 +681,54 @@ class CClientPool : public CIpPool {
+        if (m_cur_act_itr == m_active_clients.end())
+		  m_cur_act_itr = m_active_clients.begin();
+          idx = *m_cur_act_itr;
+	      m_cur_act_itr++;
```
indentation please

kvbhaskar commented on src/tuple_gen.h:
```diff
@@ -693,6 +772,12 @@ class CClientPool : public CIpPool {
+    //uint32_t m_cur_active_idx;
+
+
+    //std::vector<uint32_t> m_active_clients;
```
clean up this commented code please

kvbhaskar commented on src/tuple_gen.h:
```diff
@@ -923,6 +1013,8 @@ class CTupleGeneratorSmart {
 
     std::vector<CClientPool*> m_client_pool;
     std::vector<CServerPoolBase*> m_server_pool;
+    //std::map<uint32_t, CClientPool*>  m_ip_start_cpool_link;
```
clean up!

kvbhaskar commented on src/tuple_gen.cpp:
```diff
@@ -225,9 +287,40 @@ bool CTupleGeneratorSmart::add_client_pool(IP_DIST_t      client_dist,
+    else
+	    return NULL;
+
+    //return ((ip >= all_clients.front()->get_ip()) && (ip <= all_clients.back()->get_ip())) ? cpool : NULL;
```
cleanup please

hhaim commented:
There are still places with tabs, could you go over the code and clean this up. This is the same place I've reported.
![image](https://user-images.githubusercontent.com/13034429/93849221-44fd8680-fcb4-11ea-8e8a-7c6bb789534d.png)
Please review my comments again

hhaim commented:
@srakula-trex @kvbhaskar 
I have started to look into my comments and some was fixed but some was ignored. I stopped in comment 6.
Could you review the comments again and appreciate your comments back for each, so we can go forward 

1. ```git diff  master..srakula-trex-srakula-trex-flowtable  | grep  -P '\t'*``` shows 41 instances of tabs in the new diff
1.1 some parts of the code does not follow the guidelines 

![image](https://user-images.githubusercontent.com/13034429/93980988-0507d380-fd88-11ea-9364-870db7f83035.png)

should be

```
CTcpPerThreadCtx * ctx = pctx->m_ctx;

    if (ftp->is_tunnel_aware() && ftp->is_gtpu_tunnel()){
       /*Check if we need to add any error counter for the special case*/
         if (ctx->is_client_side()){
            INC_STAT(pctx, tp->m_flow->m_tg_id, tcps_notunnel);
         }
     }

    pkt.m_buf->dynfield1[0] = (uint64_t)ftp->m_tun_handle;   <<<<<
```

BTW this line was mentioned as a comment in **6.3** and wasn't addressed ```pkt.m_buf->dynfield1[0] = (uint64_t)ftp->m_tun_handle; ``` 


2. comment 3
>>> 3. Didn't get this code, what do you want to do here, do you want to enable it only on one port? e.g. 1 what would happen in case of multi-ports (8 ports) on which port GTPU would be enabled?
```
 for (i=0; i<port_size; i++){
            cg->m_mac_info[i].copy_src(( char *)g_opts->m_mac_addr[i].u.m_mac.src)   ;
            cg->m_mac_info[i].copy_dest(( char *)g_opts->m_mac_addr[i].u.m_mac.dest)  ;
            g_opts->m_mac_addr[i].u.m_mac.is_set = 1;
            if ( CGlobalInfo::m_options.m_enable_gtpu == i)
            {
              g_opts->m_ip_cfg[i].enable_gtp(1);
            }
```

6. comment 6
>> 6.This will not work in the common case ( no gtpu case) as the default CGlobalInfo::m_options.m_enable_gtpu == 0xFF
    if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() || CGlobalInfo::m_options.m_enable_gtpu == 0xFF)){
        /* we can't generate a flow, there is a limit*/
        return;
    }
which is now this (question remain) 
```
if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() || !(CGlobalInfo::m_options.is_gtpu_enabled()))){
        /* we can't generate a flow, there is a limit*/
        return;
    }

```

srakula-trex commented:
1.xx Removed the spaces .
2.xx  Renamed the method from set_gtpu_info_for_clients to set_tunnel_info_for_clients
3.xx  As explained over chat , we are trying to configure the ports in GTPu mode ( right now the support is only for one port on ingress to support GTPU . So it checks for the port number from the RPC ( --gtpu <port num> ) , to check which port needs to be enabled for GTPU encap .
4.xx Removed the STF code 
5.xx Removed the type from uin16 to uint8 .. 
6.xx  .  Let me go through the condition here.
New statement here is :
   if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() || !(CGlobalInfo::m_options.is_gtpu_enabled())))

This is what we do , 
	a. Check the limit ( legacy code) 
	b.  If there are no active client and gtpu is enabled .. we cannot generate flow
	c. If there are active clients, and gtpu is enabled .. we should be able to generate flow

And regression was happy , so I guess this change does work in common case as well .

7.xx	& 2.xx methods is_tunnel_aware, is_gtpu_tunnel have been added to CFlowTemplate class . And thats how we are using them now.

8.xx  : updated the case and moved that one line of copying into mbuf , also inside the Tunnel check

9.xx void *tun_handle, instead of adding it do alloc_flow better to add a function to set it and only in case of tunnel to do so  >> Differing this comment

srakula-trex commented:
@hhaim , Please check and let me know if you are OK with resolutions .

hhaim commented:
@srakula-trex did you push another commit? Did you fix the tabs issues?
See my comments from today

srakula-trex commented:
@hhaim , there was some issue in push . Pushed it again. Yes, i believe i have fixed all those tab issues now .. fingers crossed :-)

hhaim commented:
>>1.xx Removed the spaces .

1. I still see 16 instances of tabs, this is one example
![image](https://user-images.githubusercontent.com/13034429/94113482-a017b080-fe4f-11ea-9e4b-c918ccf66192.png)


```git diff  master..srakula-trex-srakula-trex-flowtable > a.txt```
and open look for re```^\+\t``` or ```^\+.*\t.*$``` 


>> 2.xx Renamed the method from set_gtpu_info_for_clients to set_tunnel_info_for_clients

good.

>> 3.xx As explained over chat , we are trying to configure the ports in GTPu mode ( right now the support is only for one port on ingress to support GTPU . So it checks for the port number from the RPC ( --gtpu ) , to check which port needs to be enabled for GTPU encap .

ok, worth adding a doc on this 

>>4.xx Removed the STF code

ok

>>5.xx Removed the type from uin16 to uint8 ..

ok

>>6.xx . Let me go through the condition here.
New statement here is :
```if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() || !(CGlobalInfo::m_options.is_gtpu_enabled())))```

ok, it wasn't clear, it seems as (```if x || y || z```) instead it is ``` if x || !(y || !z)``` that could be more clear ```if x || (z && y) ``` less not of not of not 

>>7,8

ok

>> 9.xx void *tun_handle, instead of adding it do alloc_flow better to add a function to set it and only in case of tunnel to do so >> Differing this comment

ok

we have a doc section for how to use. 

so please fix 1 and 6

hhaim commented:
In regards to **6**, I would do this:
``` if ( cur->check_limit() || !cur->m_tuple_gen.has_active_clients()  )``` there is no need to check if there is a tunnel of not

srakula-trex commented:
Fixed all the indentations  & tab settings.. hopefully it will be the last iteration for this review :-)

For 6,  The condition was kept specifically to cover the scenario where in even if you enable the GTPU and doesnt have active clients .. we don't want to generate the flows .Hence it was added .

Let me know if you have any concerns on this.

hhaim commented:
@srakula-trex in regards to 6
There is no need to ask if there is a tunnel or not, in any case that there are no client active we can’t generate flows.
With the new RPC it could happen **without** tunnel too, so there is a need to remove the last if

Rofflarn commented:
@srakula-trex how to use this in ASTF? Any example commands? I see that "--gtpu" is added, but do not see how client and server is connected here.

hhaim commented:
@Rofflarn we are in the middle of merge once it will be there, there would be a documentation . 
The GTP-C is not part of EMU

hhaim commented:
@srakula-trex 
I've changed this line and 
has_active_clients() return false when there are active clients. This code should not be related to the state of ```CGlobalInfo::m_options.is_gtpu_enabled()``` 


```
    if (cur->check_limit() || !(cur->m_tuple_gen.has_active_clients() )){
        /* we can't generate a flow, there is a limit*/
        return;
    }
```

hhaim commented:
@srakula-trex I see the issue. You are maintaining the active-clients only if RPC is called. The active clients should be maintained (list,active etc) should be always valid. consider this as a new baseline feature. Please review the the code again

hhaim commented:
rebased, need more cleanup here 
https://github.com/cisco-system-traffic-generator/trex-core/issues/532
